File: Algorithm/BOJ/C++/Dynamic Programming/11052.cpp
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

int getMaxPrice(int N, vector<int> *cards);

int main() {
	int N, price;
	vector<int> cards;

	cin >> N;
	
	cards.push_back(0);
	for (int i = 0; i < N; i++) {
		cin >> price;
		cards.push_back(price);
	}

	cout << getMaxPrice(N, &cards);
}

int getMaxPrice(int N, vector<int>* cards) {
	vector<int> answer(N + 1);

	for (int i = 1; i <= N; i++) {
		for (int j = 1; j <= i; j++) {
			answer[i] = max(answer[i], answer[i - j] + (*cards)[j]);
		}
	}

	return answer[N];
}

File: Algorithm/BOJ/Python/Dynamic Programming/11052.py
def get_max_price(N: int, cards: list) -> int:
    answer = [0 for _ in range(N + 1)]

    for i in range(1, N + 1):
        for j in range(1, i + 1):
            answer[i] = max(answer[i], answer[i - j] + cards[j])

    return answer[N]


def main():
    N = int(input())
    cards = list(map(int, input().split()))
    cards.insert(0, 0)

    print(get_max_price(N, cards))


if __name__ == '__main__':
    main()
